src/data-model-providers/endpoint/SpanEndpointProvider.cpp

```diff
@@ -0,0 +1,140 @@
+namespace app {
+
+// Builder implementation
+SpanEndpointProvider::Builder::Builder(EndpointId id) : mEndpointId(id) {}
+
+SpanEndpointProvider::Builder & SpanEndpointProvider::Builder::SetComposition(DataModel::EndpointCompositionPattern composition)
+{
+    mComposition = composition;
+    return *this;
+}
+
+SpanEndpointProvider::Builder & SpanEndpointProvider::Builder::SetParentId(EndpointId parentId)
+{
+    mParentId = parentId;
+    return *this;
+}
+
+SpanEndpointProvider::Builder & SpanEndpointProvider::Builder::SetServerClusters(Span<ServerClusterInterface *> serverClusters)
+{
+    mServerClusters = serverClusters;
+    return *this;
+}
+
+SpanEndpointProvider::Builder & SpanEndpointProvider::Builder::SetClientClusters(Span<const ClusterId> clientClusters)
+{
+    mClientClusters = clientClusters;
+    return *this;
+}
+
+SpanEndpointProvider::Builder & SpanEndpointProvider::Builder::SetSemanticTags(Span<const SemanticTag> semanticTags)
+{
+    mSemanticTags = semanticTags;
+    return *this;
+}
+
+SpanEndpointProvider::Builder & SpanEndpointProvider::Builder::SetDeviceTypes(Span<const DataModel::DeviceTypeEntry> deviceTypes)
+{
+    mDeviceTypes = deviceTypes;
+    return *this;
+}
+
+std::pair<SpanEndpointProvider, CHIP_ERROR> SpanEndpointProvider::Builder::build()
+{
+    if (mEndpointId == kInvalidEndpointId)
+    {
+        return { SpanEndpointProvider(), CHIP_ERROR_INVALID_ARGUMENT };
+    }
+
+    for (auto * cluster : mServerClusters)
+    {
+        if (cluster == nullptr || cluster->GetPaths().empty())
+        {
+            ChipLogError(DataManagement, "Builder: Attempted to build with an invalid server cluster entry.");
+            return { SpanEndpointProvider(), CHIP_ERROR_INVALID_ARGUMENT };
+        }
+    }
+
+    return { SpanEndpointProvider(mEndpointId, mComposition, mParentId, mServerClusters, mClientClusters, mSemanticTags,
+                                  mDeviceTypes),
+             CHIP_NO_ERROR };
+}
+
+// SpanEndpointProvider implementation
+SpanEndpointProvider::SpanEndpointProvider() :
+    mEndpointEntry({ kInvalidEndpointId, kInvalidEndpointId, DataModel::EndpointCompositionPattern::kFullFamily })
+// Other members are default-initialzed (empty spans)
+{}
+
+CHIP_ERROR
+SpanEndpointProvider::SemanticTags(
+    ReadOnlyBufferBuilder<chip::app::Clusters::Descriptor::Structs::SemanticTagStruct::Type> & out) const
+{
+    return out.ReferenceExisting(mSemanticTags);
+}
+
+CHIP_ERROR SpanEndpointProvider::DeviceTypes(ReadOnlyBufferBuilder<DataModel::DeviceTypeEntry> & out) const
+{
+    return out.ReferenceExisting(mDeviceTypes);
+}
+
+CHIP_ERROR SpanEndpointProvider::ClientClusters(ReadOnlyBufferBuilder<ClusterId> & out) const
+{
+    return out.ReferenceExisting(mClientClusters);
+}
+
+ServerClusterInterface * SpanEndpointProvider::GetServerCluster(ClusterId clusterId) const
+{
+    for (auto * serverCluster : mServerClusters)
+    {
+        // Check serverCluster is not null before dereferencing
```

Comment:
This comment seems to not help in reading code: it just re-states what happens below.

Can we rephrase it to explain why a serverCluster may be nullptr? Why are we not checking this when we build the provider?